relay-server/src/actors/upstream.rs

```diff
@@ -881,69 +890,115 @@ impl Entry {
 /// is synchronous and managed by the [`UpstreamBroker`].
 #[derive(Debug)]
 struct UpstreamQueue {
+    /// High priority queue.
     high: VecDeque<Entry>,
+    /// Low priority queue.
     low: VecDeque<Entry>,
+    /// Retry queue, sorted by [priority][`RequestPriority`].
+    retries: VecDeque<Entry>,
+    /// Backoff handler for retry entries.
+    retry_backoff: RetryBackoff,
+    /// Retries should not be dequeued before this instant.
+    next_retry: Instant,
 }
 
 impl UpstreamQueue {
     /// Creates an empty upstream queue.
-    pub fn new() -> Self {
+    pub fn new(retry_backoff_duration: Duration) -> Self {
         Self {
             high: VecDeque::new(),
             low: VecDeque::new(),
+            retries: VecDeque::new(),
+            retry_backoff: RetryBackoff::new(retry_backoff_duration),
+            next_retry: Instant::now(),
         }
     }
 
     /// Returns the number of entries in the queue.
     pub fn len(&self) -> usize {
-        self.high.len() + self.low.len()
+        self.high.len() + self.low.len() + self.retries.len()
     }
 
-    /// Puts an entry into the queue at the given position.
-    fn put(&mut self, entry: Entry, position: EnqueuePosition) {
+    /// Puts an entry into the queue.
+    ///
+    /// High priority entries go in the front of the queue, and low priority
+    /// items in the back. [Retries][RequestAttempt::Retry] are tracked
+    /// independently.
+    fn put(&mut self, entry: Entry, attempt: RequestAttempt) {
```

Comment:
This function no longer needs to exist. `enqueue` and `enqueue_retry` do entirely separate things, so inline their respective implementations and make them look as similar as possible.